### ISSUE TITLE
Slicing inefficient for larger signals

### ISSUE THREAD

**Issue by studywolf** [nengo/nengo#386]
So I was sending both neuron activities and a signal into a node, and then slicing out the signal (shown in ```out_fun1```, using ```node c```) to get my activities and my learning signal and then returning the dot product, and when I run this it takes about 5.6 seconds. 
Terry suggested going about it in the other way, using ```out_fun2``` and ```store_func```, with ```d``` and ```e```, and my runtime is 2.3 seconds. So it could be just that it's a numpy thing or maybe there's something going on like a giant identity matrix multiply somewhere that's maybe possible to workaround, but wanted to throw it out there!
```
import numpy as np
import nengo
nengo.log()

m = nengo.Network(seed=10)

def out_fun1(t,x):
    sig1 = x[:3]
    sig2 = x[3:]
    return -.01 * np.outer(sig1, sig2).flatten()

activities = np.zeros(500)
def out_fun2(t,x):
    return -.01 * np.outer(x, activities).flatten()
def store_func(t,x):
    global activities
    activities = x

with m: 

    a = nengo.Ensemble(n_neurons=100, dimensions=3)
    b = nengo.Ensemble(n_neurons=500, dimensions=1)
    # c = nengo.Node(output=out_fun1, size_in=3+500)
    d = nengo.Node(output=out_fun2, size_in=3)
    e = nengo.Node(output=store_func, size_in=500)

    # nengo.Connection(a, c[:3])
    # nengo.Connection(b.neurons, c[3:])
    nengo.Connection(a, d)
    nengo.Connection(b.neurons, e)

sim = nengo.Simulator(m)
import time
start = time.time()
sim.run(10)
print time.time() - start
```
 
studywolf commented on May 13
@hunse! @arvoelke and I were talking about it and he mentioned that it's likely just the second big transform in the case when you're using c that's causing the slow down, but it might be possible to do something like compress those two connections created to c[:3] and c[3:] into one connection. Do you have any thoughts on that? 
Alternatively we could just document that it's more efficient to kind of hack around it, but it's a bit of shame because it seems like the c way is more intuitive, with both signals projecting into a single node.
 
hunse commented on May 20
There's no easy way to compress the two connections into one, since the resulting connection would have two sources. What would be easier is to not make the full transform all the time, and instead allow connections that map from slices of objects to slices of objects. This shouldn't be too hard to do, but it's definitely a wishlist issue.

**Comment by jgosmann** [comment]
Output on my computer:
```
ctn18:~> python test.py
Simulation finished in 0:00:02.                                                                                         
1.6690633296966553
```

**Comment by tbekolay** [comment]
Output as posted:

```
Simulation finished in 0:00:02.                                                 
1.59112286568
```

With slow lines:

```
Simulation finished in 0:00:02.                                                 
1.72944402695
```

So basically the same results as @jgosmann. I think it's reasonable to expect that slicing will add a bit of time, so it's likely that something we've done in the past few years has made this operation faster.

It could be possible that this is an issue that crops up depending on the NumPy installation... so to see if that's the case, I checked out an older version of Nengo to see if anything's changed.

Here's what I got with Nengo v2.0.0 with the fast lines:

```
Simulation finished in 0:00:02.                                                 
1.27331900597
```

and the slow lines:

```
Simulation finished in 0:00:03.                                                 
2.7261030674
```

My computer was also noticeably working harder with the slower lines, so clearly something has changed such that the issue @studywolf had has been fixed. The fact that we're now running slightly slower than before is a bit concerning though!
